### ISSUE TITLE
Support for files endpoint

### ISSUE THREAD

**Issue by ml-evs** [materialscloud-org/optimade-maker#32]
We could consider adding support for the OPTIMADE v1.2 files endpoint, which would allow each structure to be linked to additional files that may not follow the OPTIMADE format. This hinges on whether the eventually release v1.2 format will be flexible enough to be able to reference files within archives.

**Comment by Bud-Macaulay** [comment]
Hi! @eimrek and @ml-evs, was wondering where we stood on this - i think i have a related issue?

I am currently investigating migrating whether some of our backend could be fully derived through optimade and since we have a wide range of attached properties (aswell as access to large data storage) it makes sense to serve files through optimade-python-tools. 

_Note that often properites are often too big to store inside a given structure endpoint without severally affecting performance_. 

For now i am investigating whether a specific tool hosted on materialscloud can be ported (topomat), which is currently running a custom postgres backend.

In particular for a given structure there is some meta-data, properties and images that i would like associated.
cifs/
  ${formula}_${uuid}.cif # CIF
json/
  ${formula}_${uuid}.json # METADATA [Likely small enough to attach directly to the structure ]
wcc/
 ${formula}_${uuid}.png 
files/
 ${formula}_${uuid}_${calculationN}-out.txt 
 ${formula}_${uuid}_${calculationN}-in.txt

I would like to have for a given structure response, attached relationships to files so a consumer can easily inspect and grab meta data as and when needed.

I note this block in the source mongo_utils:
``` python 
                if "relationships" in entry:
                    inp_data["relationships"] = entry["relationships"]
                if "links" in entry:
                    inp_data["links"] = entry["links"]
```
perhaps we can add relationships (and files) to the optimade.yaml - (if its not already there)

**Comment by ml-evs** [comment]
From the optimade-python-tools side, its just waiting for someone (me?) to finish off Johan's initial work at https://github.com/Materials-Consortia/optimade-python-tools/pull/1414 -- might be a good target for the upcoming workshop. The code snippet you posted I think shows that if you provide a preprocessed JSONL file then it will not throwaway the relationships/links, there might be some validation complaints that o-p-t doesn't know what a file is, but we could try to push through those. The main sticking point with that original PR is that we had no nice way for people to actually put files in the db and serve them from somewhere, maybe optimade-maker actually helps solve that, so we could get away with just adding the models/endpoints.

**Comment by Bud-Macaulay** [comment]
For my usecase, its highly likely that the files I would like to reference will be completely separate from the optimade deployment and instead referenced only by URI/URL (being served through a public bucket).

Therefore any functionality that allows me to make relationships that point to this bucket through the optimade.yaml file rather than having to link  them manually in the resulting jsonl would be great and I think this wouldn't require too much hacking. 

However, i do note that perhaps a self-contained solution that allows linking and serving files together is the more flexible usecase.

**Comment by ml-evs** [comment]
Right, understood -- the files endpoint would be ideal for this (its just a listing of file metadata with links out). You could try populating your local db with OPTIMADE-file-like things and then try to run the optimade-python-tools version in that PR (it will be a little old) to see if that does the trick? Then here we would just need to wait for the optimade-python-tools models to catch-up and it should already work when going direct from JSONL. 

Would you need the config to be extended to include files and allow you to match files to structures when creating a JSONL yourself?

**Comment by Bud-Macaulay** [comment]
> Would you need the config to be extended to include files and allow you to match files to structures when creating a JSONL yourself?

Personally, no. I can manage to write a small script that just attaches files to structures (and build the URLs from folder structure).

Maybe for other users attaching files through the config would be useful however if they aren't currently served perhaps it should wait before being implemented here.
